src/node/internal/internal_http_server.ts

```diff
@@ -0,0 +1,730 @@
+import {
+  kOutHeaders,
+  WrittenDataBufferEntry,
+  HeadersSentEvent,
+} from 'node-internal:internal_http_outgoing';
+import {
+  chunkExpression,
+  _checkInvalidHeaderChar,
+} from 'node-internal:internal_http';
+import { _normalizeArgs } from 'node-internal:internal_net';
+import { Buffer } from 'node-internal:internal_buffer';
+
+import type {
+  Server as _Server,
+  ServerResponse as _ServerResponse,
+  RequestListener,
+  ServerOptions,
+  OutgoingHttpHeaders,
+  OutgoingHttpHeader,
+} from 'node:http';
+import type { Socket, AddressInfo } from 'node:net';
+import { default as flags } from 'workerd:compatibility-flags';
+
+export const kConnectionsCheckingInterval = Symbol(
+  'http.server.connectionsCheckingInterval'
+);
+
+export type DataWrittenEvent = {
+  index: number;
+  entry: WrittenDataBufferEntry;
+};
+
+export class Server
+  extends EventEmitter
+  implements _Server, BaseWithHttpOptions
+{
+  // @ts-expect-error TS2416 Server is not assignable to same property in base type.
+  [kIncomingMessage]: typeof IncomingMessage = IncomingMessage;
+  // @ts-expect-error TS2416 Server is not assignable to same property in base type.
+  [kServerResponse]: typeof ServerResponse = ServerResponse;
+  [kConnectionsCheckingInterval]?: number;
+  [kUniqueHeaders]: Set<string> | null = null;
+
+  // Similar option to this. Too lazy to write my own docs.
+  // http://www.squid-cache.org/Doc/config/half_closed_clients/
+  // https://wiki.squid-cache.org/SquidFaq/InnerWorkings#What_is_a_half-closed_filedescriptor.3F
+  httpAllowHalfOpen = false;
+  timeout = 0;
+  maxHeadersCount: number | null = null;
+  maxRequestsPerSocket = 0;
+  connectionsCheckingInterval = 30_000;
+  requestTimeout: number = 0;
+  headersTimeout: number = 0;
+  requireHostHeader: boolean = false;
+  joinDuplicateHeaders: boolean = false;
+  rejectNonStandardBodyWrites: boolean = false;
+  keepAliveTimeout: number = 5_000;
+  #port: number | null = null;
+
+  constructor(options?: ServerOptions, requestListener?: RequestListener) {
+    if (!flags.enableNodejsHttpServerModules) {
+      throw new ERR_METHOD_NOT_IMPLEMENTED('Server');
+    }
+    super();
+
+    if (options != null) {
+      // @ts-expect-error TS2345 TODO(soon): Find a better way to handle this type mismatch.
+      storeHTTPOptions.call(this, options);
+    }
+
+    // TODO(soon): Support options.highWaterMark option.
+
+    if (typeof options === 'function') {
+      requestListener = options;
+      options = {};
+    } else if (options == null) {
+      options = {};
+    } else {
+      validateObject(options, 'options');
+    }
+
+    if (requestListener) {
+      this.on('request', requestListener);
+    }
+
+    this[kUniqueHeaders] = parseUniqueHeadersOption(
+      options.uniqueHeaders as (string | string[])[]
+    );
+  }
+
+  close(callback?: VoidFunction): this {
+    httpServerPreClose(this);
+    if (this.#port != null) {
+      portMapper.delete(this.#port);
```

Comment:
Since the FinalizationRegistry stuff was removed, it would be good to add a comment here and in the documentation that failing to call close may result in the server being leaked. I think also, just in general, in the documentation for all this, we need to warn folks away from calling `createServer(...)` within a request to avoid the possibility of bugs creating and leaking server instances.